baselines/hfedxgboost/hfedxgboost/client.py

```diff
@@ -0,0 +1,336 @@
+"""Define your client class and a function to construct such clients.
+
+Please overwrite `flwr.client.NumPyClient` or `flwr.client.Client` and create a function
+to instantiate your client.
+"""
+from typing import Tuple, Union
+
+import flwr as fl
+import torch
+import torch.nn as nn
+from flwr.common import (
+    Code,
+    EvaluateIns,
+    EvaluateRes,
+    FitIns,
+    FitRes,
+    GetParametersIns,
+    GetParametersRes,
+    Status,
+    ndarrays_to_parameters,
+    parameters_to_ndarrays,
+)
+from torch.utils.data import DataLoader
+from torchmetrics import Accuracy, MeanSquaredError
+from tqdm import tqdm
+from xgboost import XGBClassifier, XGBRegressor
+
+from hfedxgboost.models import CNN, fit_XGBoost
+from hfedxgboost.utils import single_tree_preds_from_each_client
+
+
+class FL_Client(fl.client.Client):
+    """Custom class contains the methods that the client need."""
+
+    def __init__(
+        self,
+        cfg,
+        trainloader: DataLoader,
+        valloader: DataLoader,
+        client_num: int,
+        cid: str,
+        log_progress: bool = False,
+    ):
+        self.task_type = cfg.dataset.task.task_type
+        self.cid = cid
+        self.config = cfg
+        for dataset in trainloader:
+            data, label = dataset[0], dataset[1]
+        self.tree = fit_XGBoost(cfg, self.task_type, data, label, 100)
+
+        self.trainloader_original = trainloader
+        self.valloader_original = valloader
+        self.trainloader = None
+        self.valloader = None
+        self.n_estimators_client = (
+            cfg.n_estimators_client
+        )  # 100#cfg.dataset.n_estimators_client
+        self.client_num = client_num
+        self.properties = {"tensor_type": "numpy.ndarray"}
+        self.log_progress = log_progress
+
+        # instantiate model
+        self.net = CNN(cfg)
+
+        # determine device
+        self.device = (
+            "cpu"  # torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        )
```

Comment:
```suggestion
        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```

Reply:
If you pass `num_gpus=0` to your `client_resources` in `start_simulation()`, this will always be evaluated as `"cpu"`. It will use gpu only if `num_gpus>0`.

If there is a reason why you found it necessary to be hardcoded to "cpu"?

Reply:
I just love to use my CPU as I have a weak integrated GPU

Reply:
but the above will always use CPU if your `num_gpus` in `client_resources` is set to 0. My suggestion is to not hardcode it other wise if someone else wants to later use GPU, raising `num_gpus>0` won't do anything.

Reply:
oh, I get it, I've already committed that change